src/mem/cache/tags/partitioning_policies/max_capacity.cc

```diff
@@ -0,0 +1,114 @@
+ * not be construed as granting a license to any other intellectual
+ * property including but not limited to intellectual property relating
+ * to a hardware implementation of the functionality of the software
+ * licensed hereunder.  You may use the software subject to the license
+ * terms below provided that you ensure that this notice is replicated
+ * unmodified and in its entirety in all distributions of the software,
+ * modified or unmodified, in source code or in binary form.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met: redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer;
+ * redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "mem/cache/tags/partitioning_policies/max_capacity.hh"
+
+#include <algorithm>
+#include <string>
+
+#include "params/MaxCapacityPartitioningPolicy.hh"
+
+namespace gem5
+{
+
+namespace partitioning_policy
+{
+
+MaxCapacity::MaxCapacity(const MaxCapacityPartitioningPolicyParams &params)
+    : Base(params), cacheSize(params.cache_size), blkSize(params.blk_size)
+{
+    for (const auto &cfg_s : params.partitionId_max_capacity) {
+        const auto first_delim = cfg_s.find(',');
+        const uint64_t partition_id = std::stoul(cfg_s.substr(0, first_delim));
+        const double cap_frac = std::stod(cfg_s.substr(first_delim+1));
+
+        // check Capacity Fraction (cap_frac) is actually a fraction in [0,1]
+        if (!(cap_frac>=0 && cap_frac<=1))
+            panic("MaxCapacity Partitioning Policy for partition_ID %d has \
+            Capacity Fraction %f outside of [0,1] range",
+            partition_id, cap_frac);
```

Comment:
A many line if statement needs `{` curly braces. Too many times, we've had bugs in gem5 because of skipping the curly brace in cases like these.